src/lib/ewelink.ts

```diff
@@ -1,6 +1,697 @@
-import {presets} from './exposes';
-import {battery, setupConfigureForBinding} from './modernExtend';
-import {Configure, Expose, Fz, KeyValueAny, ModernExtend} from './types';
+// import fz from '../converters/fromZigbee';
+
+import tz from '../converters/toZigbee';
+import {access, Cover, options, presets} from './exposes';
+import {battery, setupConfigureForBinding, setupConfigureForReporting} from './modernExtend';
+import {Configure, DefinitionMeta, Expose, Fz, KeyValueAny, ModernExtend, Tz, Zh} from './types';
+import * as utils from './utils';
+
+const e = presets;
+const ea = access;
+
+// ====================== Type Or Interface ==============================
+interface WindowCoveringArgs {
+    controls: ('lift' | 'tilt')[];
+    coverInverted?: boolean;
+    stateSource?: 'lift' | 'tilt';
+    configureReporting?: boolean;
+    coverMode?: boolean;
+    endpointNames?: string[];
+}
+
+// ======================= Utils =========================================
+const findKeyByValue = (object: object, value: number | string) => {
+    const entry = Object.entries(object).find(([key, val]) => val === value);
+    return entry ? entry[0] : undefined;
+};
+// ======================= Custom TZ =====================================
+export const ewelinkToZigbee = {
+    cover_state: {
+        key: ['state'],
+        convertSet: async (entity, key, value, meta) => {
+            const lookup = {open: 'upOpen', close: 'downClose', stop: 'stop', on: 'upOpen', off: 'downClose'};
+            utils.assertString(value, key);
+            value = value.toLowerCase();
+            await entity.command('closuresWindowCovering', utils.getFromLookup(value, lookup), {}, utils.getOptions(meta.mapped, entity));
+            return {
+                state: {
+                    [key]:
+                        value === 'on'
+                            ? 'open'.toLocaleUpperCase()
+                            : value === 'off'
+                              ? 'close'.toLocaleUpperCase()
+                              : (value as string).toLocaleUpperCase(),
+                },
+            };
+        },
+    } satisfies Tz.Converter,
+};
+
+// ======================= Custom FZ =====================================
+export const ewelinkFromZigbee = {
+    cover_position_tilt: {
```

Comment:
Could you also prevent code duplication here?

Reply:
Thank you for the review. We will modify the code, complete the testing, and resubmit it